In [1]:
!pip install -q pyomo

     |████████████████████████████████| 8.9 MB 6.0 MB/s 
     |████████████████████████████████| 49 kB 5.9 MB/s 


In [2]:
from pyomo.environ import *

## **Que.1**


In [3]:
model1= ConcreteModel()

In [4]:
#declare decision variable
model1.x1=Var(domain=Reals)
model1.x2=Var(domain=Reals)
model1.x3=Var(domain=Reals)
model1.x4=Var(domain=Reals)
model1.x5=Var(domain=Reals)
model1.x6=Var(domain=Reals)
model1.x7=Var(domain=Reals)

In [5]:
#declare objective
model1.cost=Objective(expr=2*model1.x1-5*model1.x2+14*model1.x3+7*model1.x4-11*model1.x5+0.17*model1.x6-10.2*model1.x7,sense=minimize)

In [6]:
#declare constraint
model1.constr1=Constraint(expr=model1.x1+model1.x2+model1.x3+model1.x4+model1.x5+model1.x6+model1.x7==280)
model1.constr2=Constraint(expr=model1.x1-1*model1.x2>=10)
model1.constr3=Constraint(expr=-3*model1.x3+model1.x6-3*model1.x7<=205)
model1.constr4=Constraint(expr=20*model1.x1+42*model1.x6+model1.x5>=35)
model1.constr5=Constraint(expr=model1.x2-model1.x3+4*model1.x7>=24)
model1.constr6=Constraint(expr=model1.x1>=12)
model1.constr7=Constraint(expr=model1.x2>=0)
model1.constr8=Constraint(expr=model1.x3>=-5)
model1.constr9=Constraint(expr=model1.x4>=3)
model1.constr10=Constraint(expr=model1.x5>=10)
model1.constr11=Constraint(expr=model1.x6>=-4)
model1.constr12=Constraint(expr=model1.x7>=0)


In [7]:
model1.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x7 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fi

In [8]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 160837 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [9]:
SolverFactory('glpk',executable='/usr/bin/glpsol').solve(model1).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -3035.88
  Upper bound: -3035.88
  Number of objectives: 1
  Number of constraints: 13
  Number of variables: 8
  Number of nonzeros: 26
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.010588645935058594
# ----------------------------------------------------------
#   Solution Information
# ----------------------------



# Value of the optimal function and values of the decision variables:-

In [10]:
print('Objective fuction value= ',model1.cost())
print('\nvalue of X1= ',model1.x1())
print('value of X2= ',model1.x2())
print('value of X3= ',model1.x3())
print('value of X4= ',model1.x4())
print('value of X5= ',model1.x5())
print('value of X6= ',model1.x6())
print('value of X7= ',model1.x7())


Objective fuction value=  -3035.8799999999997

value of X1=  12.0
value of X2=  0.0
value of X3=  -5.0
value of X4=  3.0
value of X5=  269.25
value of X6=  -4.0
value of X7=  4.75


# Constraint Activities

In [11]:
#constraint activities
#first constraint
print("The value of first constraint=",model1.constr1())

The value of first constraint= 280.0


The value of the first constraint is 280.0 which is equal to RHS (i.e.280)

In [12]:
#second constraint
print("The value of second constraint = ",model1.constr2())

The value of second constraint =  12.0


The value of the second constraint is 12.0 which is greater than RHS(i.e.10)

In [13]:
#third constraint
print("The value of Third constraint = ",model1.constr3())

The value of Third constraint =  -3.25


The value of third constraint is -3.25 which is lesser than RHS(i.e. 205)

In [14]:
#fourth constraint
print("The value of fourth constraint = ",model1.constr4())

The value of fourth constraint =  341.25


The value of the fourth constraint is 341.25 which is greater than RHS (i.e.35)

## **Que.2**

$\max\ 0.043x_1+0.027x_2+0.025x_3+0.022x_4+0.045x_5-0.0275y \\\text{s.t.}\\ x_1+3x_2+x_3+7x_4+x_5-y\leq 100\\x_2+2x_3+4x_4\geq 60 \\ 0.8x_1-4.8x_2+1.8x_3-3.8x_4-2.2x_5\geq 0\\ x_1-12x_2-x_3-3x_4+5x_5\geq 0 \\ x_1,x_2,x_3,x_4,x_5 \geq 0 \\ y\geq0\\ y\leq2.25$


**Assumptions:-**


$\\
1.x_1-x_2+x_3-x_4+x_5 \geq 0 \\ 2. x_1+x_2+x_3-x_5 \geq 0$

In [15]:
model2= ConcreteModel()

In [16]:
#declare decision variable
model2.x1=Var(domain=Reals)
model2.x2=Var(domain=Reals)
model2.x3=Var(domain=Reals)
model2.x4=Var(domain=Reals)
model2.x5=Var(domain=Reals)
model2.y=Var(domain=Reals)

In [17]:
#declare Objective
model2.cost=Objective(expr=0.043*model2.x1+0.027*model2.x2+0.025*model2.x3+0.022*model2.x4+0.045*model2.x5-0.0275*model2.y,sense=maximize)

In [18]:
#declare constraint
model2.constr1=Constraint(expr=model2.x1+3*model2.x2+model2.x3+7*model2.x4+model2.x5-model2.y<=100)
model2.constr2=Constraint(expr=model2.x2+2*model2.x3+4*model2.x4>=60)
model2.constr3=Constraint(expr=(0.8)*model2.x1-(4.8)*model2.x2+(1.8)*model2.x3-(3.8)*model2.x4-(2.2)*model2.x5>=0)
model2.constr4=Constraint(expr=model2.x1-12*model2.x2-model2.x3-3*model2.x4+5*model2.x5>=0)
model2.constr5=Constraint(expr=model2.x1>=0)
model2.constr6=Constraint(expr=model2.x2>=0)
model2.constr7=Constraint(expr=model2.x3>=0)
model2.constr8=Constraint(expr=model2.x4>=0)
model2.constr9=Constraint(expr=model2.x5>=0)
model2.constr10=Constraint(expr=model2.y>=0)
model2.constr11=Constraint(expr=model2.y<=2.25)

In [19]:
model2.pprint()

6 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
 

In [20]:
SolverFactory('glpk',executable='/usr/bin/glpsol').solve(model2).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 3.86940833333333
  Upper bound: 3.86940833333333
  Number of objectives: 1
  Number of constraints: 12
  Number of variables: 7
  Number of nonzeros: 27
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.0193326473236084
# ----------------------------------------------------------
#   Solution Information
# --------------

# value of the optimal function and the values of the decision variable:-

In [21]:
print("The optimal solution value is = ",model2.cost())
print("\nThe value of the variable x1= ",model2.x1())
print("The value of the variable x2= ",model2.x2())
print("The value of the variable x3= ",model2.x3())
print("The value of the variable x4= ",model2.x4())
print("The value of the variable x5= ",model2.x5())
print("The value of the variable y= ",model2.y())


The optimal solution value is =  3.869408333333333

The value of the variable x1=  34.9833333333333
The value of the variable x2=  0.0
The value of the variable x3=  30.0
The value of the variable x4=  0.0
The value of the variable x5=  37.2666666666667
The value of the variable y=  2.25


# Constraint Activities

In [22]:
#constraint activity
#first constraint
print("The value of the first constraint= ",model2.x1()+3*model2.x2()+model2.x3()+7*model2.x4()-model2.x5()-model2.y())

The value of the first constraint=  25.46666666666659


The value of first condtraint is 25.46666666666659 which is lesser than RHS (i.e.100)

In [23]:
#second Constraint
print("The value of the second constraint= ",model2.x2()+2*model2.x3()+4*model2.x4())

The value of the second constraint=  60.0


The value of the second constraint is 60 which is equal to the RHS (i.e.60)

In [24]:
#fourth constraint
print("The value of the fourth constraint= ",model2.constr4())

The value of the fourth constraint=  191.3166666666668


The value of the fourth constraint is 191.3166666666668 which is greater than RHS(i.e.0)

In [25]:
print("The value of fifth constraint =",model2.constr5())

The value of fifth constraint = 34.9833333333333


The value of fifth constraint is 34.9833333333333 which is greater than RHS(i.e.0)

In [26]:
print("The value of the sixth constraint=",model2.constr6())

The value of the sixth constraint= 0.0


The value of sixth constraint is 0 which is equal to RHS(i.e.0)

In [27]:
print("The value of seventh constraint is=",model2.constr7())

The value of seventh constraint is= 30.0


The value of 7th constraint is 30 which is greater than RHS(i.e.0)

In [28]:
print("The value of 8th constraint=",model2.constr8())

The value of 8th constraint= 0.0


The value of 8th constraint is 0 which is equal to RHS(i.e.0)

In [29]:
print("The value of 9th constraint= ",model2.constr9())

The value of 9th constraint=  37.2666666666667


The value of 9th constraint is 37.2666666666667 which is greater than RHS(i.e.0)

In [30]:
print("The value of the 10th constraint is= ",model2.constr10())

The value of the 10th constraint is=  2.25


The value of the 10th constraint is 2.25 which is greater than RHS(i.e.0)

In [31]:
print("The value of the 11th constraint is= ",model2.constr11())

The value of the 11th constraint is=  2.25


The value of the 11th constraint is 2.25 which is equal to the RHS(i.e.2.25)

# Modification of Lpp without any assumption


**if we try to modify the LPP without any assumption then we have to add some other constraint in lpp. so the modified LPP is as follows:-**
$\max\  0.043x_1+0.027x_2+0.025x_3+0.022x_4+0.045x_5-0.0275y \\\text{s.t.}\\ x_1+3x_2+x_3+7x_4+x_5-y\leq 100\\x_2+2x_3+4x_4\geq 60 \\ 0.8x_1-4.8x_2+1.8x_3-3.8x_4-2.2x_5\geq 0\\ x_1-12x_2-x_3-3x_4+5x_5\geq 0 \\x_1+x_2+x_3-x_5 \geq 0\\x_1-x_2+x_3-x_4+x_5 \geq 0\\ x_1,x_2,x_3,x_4,x_5 \geq 0 \\ y\geq0\\ y\leq2.25$


## **Que.03**

$\text{let x is the quantity(in kg) of the product P and y is the quantity(in kg) of the product Q}\\ 
\text{Objective:-} $
$$\text{min}\ 145x+280y$$
$\text{such that}$
$$x+y\geq 274\\ x\geq84\\ y\geq80\\69x+55y\leq25800
$$

In [32]:
model3=ConcreteModel()

In [33]:
#declare variable
model3.x=Var(domain=Reals)
model3.y=Var(domain=Reals)

In [34]:
#declare objective
model3.cost=Objective(expr=145*model3.x+280*model3.y,sense=minimize)

In [35]:
#declare constraint
model3.constr1=Constraint(expr=model3.x+model3.y>=274)
model3.constr2=Constraint(expr=model3.x>=84)
model3.constr3=Constraint(expr=model3.y>=80)
model3.constr4=Constraint(expr=69*model3.x+55*model3.y<=25800)


In [36]:
model3.pprint()

2 Var Declarations
    x : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 145*x + 280*y

4 Constraint Declarations
    constr1 : Size=1, Index=None, Active=True
        Key  : Lower : Body  : Upper : Active
        None : 274.0 : x + y :  +Inf :   True
    constr2 : Size=1, Index=None, Active=True
        Key  : Lower : Body : Upper : Active
        None :  84.0 :    x :  +Inf :   True
    constr3 : Size=1, Index=None, Active=True
        Key  : Lower : Body : Upper : Active
        None :  80.0 :    y :  +Inf :   True
    constr4 : Size=1, Index=None, Active=True
        Key  : Lower : Body     

In [37]:
SolverFactory('glpk',executable='/usr/bin/glpsol').solve(model3).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 50530.0
  Upper bound: 50530.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 3
  Number of nonzeros: 7
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.06330132484436035
# ----------------------------------------------------------
#   Solution Information
# ---------------------------------

# Value of the optimal function and values of the decision variables:-

In [38]:
print("The optimal value =",model3.cost())
print("The quantity of product P= ",model3.x())
print("The quantity of product Q= ",model3.y())

The optimal value = 50530.0
The quantity of product P=  194.0
The quantity of product Q=  80.0


# Constraint Activity

In [39]:
print("The value of first constraint= ",model3.constr1())

The value of first constraint=  274.0


The value of the first constraint is 274 which is equal to RHS(i.e.274)

In [40]:
print("The value of the second constraint= ",model3.constr2())

The value of the second constraint=  194.0


The value of the second constraint is 194 which is greater than RHS(i.e. 84)

In [41]:
print("The value of the third constraint= ",model3.constr3())

The value of the third constraint=  80.0


The value of the third constraint is 80 which is equal to RHS(i.e 80)

In [42]:
print("The value of the fourth constraint= ",model3.constr4())

The value of the fourth constraint=  17786.0


The value of the fourth constraint is 17786 which is less than RHS(i.e 25800)

# Modification in the LP model after an additional order:-

**After an additional order the formulation will change as follows:-**

**1.The minimum value of product P and product Q will increase by 102kg and 107kg respectively.**

**2.The maximum value of x+y will increase by 102+117=219.**

$\text{let x is the quantity(in kg) of the product P and y is the quantity(in kg) of the product Q}\\ 
\text{Objective:-} $
$$\text{min}\ 145x+280y$$
$\text{such that}$
$$x+y\geq 493\\ x\geq186\\ y\geq197\\69x+55y\leq25800
$$

In [43]:
model32=ConcreteModel()

In [44]:
#declare variable
model32.x=Var(domain=Reals)
model32.y=Var(domain=Reals)


In [45]:
#declare objective
model32.cost=Objective(expr=145*model32.x+280*model32.y,sense=minimize)

In [46]:
#declare constraint
model32.constr1=Constraint(expr=model32.x+model32.y>=493)
model32.constr2=Constraint(expr=model32.x>=186)
model32.constr3=Constraint(expr=model32.y>=197)
model32.constr4=Constraint(expr=69*model32.x+55*model32.y<=25800)

In [47]:
model32.pprint()

2 Var Declarations
    x : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 145*x + 280*y

4 Constraint Declarations
    constr1 : Size=1, Index=None, Active=True
        Key  : Lower : Body  : Upper : Active
        None : 493.0 : x + y :  +Inf :   True
    constr2 : Size=1, Index=None, Active=True
        Key  : Lower : Body : Upper : Active
        None : 186.0 :    x :  +Inf :   True
    constr3 : Size=1, Index=None, Active=True
        Key  : Lower : Body : Upper : Active
        None : 197.0 :    y :  +Inf :   True
    constr4 : Size=1, Index=None, Active=True
        Key  : Lower : Body     

In [48]:
SolverFactory('glpk',executable='/usr/bin/glpsol').solve(model32).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 3
  Number of nonzeros: 7
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: other
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.014882802963256836


**This linear program is unbounded,so we don't get any optimal solution.I would like to suggest the maple industries to increase the avalilable time to fulfill tha new requirement.**